cluster/directory/base_directory.go

```diff
@@ -39,13 +52,48 @@ func NewBaseDirectory(url *common.URL) BaseDirectory {
 		destroyed: atomic.NewBool(false),
 	}
 }
+func (dir *BaseDirectory) Destroyed() bool {
+	return dir.destroyed.Load()
+}
 func (dir *BaseDirectory) GetUrl() common.URL {
 	return *dir.url
 }
 func (dir *BaseDirectory) GetDirectoryUrl() *common.URL {
 	return dir.url
 }
 
+func (dir *BaseDirectory) SetRouters(routers []cluster.Router) {
+	routerKey := dir.GetUrl().GetParam(constant.ROUTER_KEY, "")
+	if len(routerKey) > 0 {
+		factory := extension.GetRouterFactory(dir.GetUrl().Protocol)
+		url := dir.GetUrl()
+		router, err := factory.Router(&url)
+		if err == nil {
+			routers = append(routers, router)
+		}
+	}
+
+	dir.mutex.Lock()
+	defer dir.mutex.Unlock()
+	dir.routers = routers
+}
+
+func (dir *BaseDirectory) Routers() []cluster.Router {
+	dir.once.Do(func() {
+		rs := routerURLSet.Values()
+		for _, r := range rs {
+			factory := extension.GetRouterFactory(r.(*common.URL).GetParam("router", "condition"))
+			router, err := factory.Router(r.(*common.URL))
+			if err != nil {
+				logger.Errorf("router fail! error:%v", err)
+				continue
+			}
+			dir.routers = append(dir.routers, router)
```

Comment:
hey, guy, u  should use dir.mutex to protect dir.routers even if in once.Do.

Reply:
this method has deleted